[api] fix hostsystem repository handling

adrianschroeter opened the pull request:
backend supports multiple entries. They are just a different kind of
path elements. So handling them via a flag in path elements instead
of heaving a single fixed entry per repository (actually per project
due to bug in former implementation).

_Replace this line with a description of your changes. Please follow the suggestions in the comment below._

-----------------

If this PR requires any particular action or consideration before deployment,
please check the reasons or add your own to the list:

* [ ] Requires a database migration that can cause downtime. Postpone deployment until maintenance window[1].
* [ ] Contains a data migration that can cause temporary inconsistency, so should be run at a specific point of time.
* [ ] Changes some configuration files (e.g. options.yml), so the changes have to be applied manually in the reference server.
* [ ] A new Feature Toggle[2] should be enabled in the reference server.
* [ ] Proper documentation or announcement has to be published upfront since the introduced changes can confuse the users.

[1] https://github.com/openSUSE/open-build-service/wiki/Deployment-of-build.opensuse.org#when-there-are-migrations
[2] https://github.com/openSUSE/open-build-service/wiki/Feature-Toggles-%28Flipper%29#you-want-real-people-to-test-your-feature

<!---
If you haven't done so already, please read the CONTRIBUTING.md file to learn
how we work and what we expect from all contributors.

https://github.com/openSUSE/open-build-service/blob/master/CONTRIBUTING.md

In order to make it as easy as possible for other developers to review your
pull request we ask you to:

- Explain what this PR is about in the description
- Explain the steps the reviewer has to follow to verify your change
- If the reviewer needs sample data to verify your change, please explain how to
  create that data
- If you include visual changes in this PR, please add screenshots or GIFs
- If you address performance in this PR, add benchmark data or explain how the
  reviewer can benchmark this

This is a good PR description example:

Hey Friends,

this introduces labels for the different build result states on the project
monitor page. This makes it easier to get a visual overview of what is going on
in your project.

To verify this feature

- Enable the interconnect to build.opensuse.org
- Create the project home:Admin
- Add 'openSUSE Tumbleweed' as a repository to the project
- Branch a couple of packages into the project:
  ```
  for i in `osc -A http://0.0.0.0:3000 ls openSUSE.org:home:hennevogel`; do osc -A http://0.0.0.0:3000 copypac openSUSE.org:home:hennevogel $i home:Admin; done
  ```
- Visit the monitor page and see the new labels for the different states.

Here is a screenshot of how it looks:

** Before **
![Screenshot of the project monitor](https://example.com/screenshot1.png)

** After **
![Screenshot of the project monitor](https://example.com/screenshot2.png)

If this PR requires any particular action or consideration before deployment,
set out the reasons by checking the items in the list or adding your own items.
-->

dmarcoux commented on src/api/db/migrate/20210603113812_fix_hostsystem_definition.rb:
```diff
@@ -0,0 +1,17 @@
+
+  def up
+    # no real production data inside yet, so we skip any conversation.
+    safety_assured { remove_column :repositories, :hostsystem_id }
```
It doesn't matter if the column has production data or not, removing a column is **not** safe. Please follow the recommendations from [strong_migrations](https://github.com/ankane/strong_migrations#removing-a-column) on how to safely drop a column. Without this, we're going to be down once this migration runs.

adrianschroeter replied:
that way?

dmarcoux replied:
1. Change the code to ignore the column you want to drop.
2. Deploy
3. Create a migration to remove the column
4. Deploy and run the migration.

This cannot happen all at the same time. It has to be done in two steps, so it will have to be in 2 PRs. It's a bit more work, but then we're not down because we remove a column.

adrianschroeter replied:
as I said, this column couldn't have used before, if it was the result would be broken anyway.

But we can move just the removal to another later migration if wanted.

dmarcoux replied:
It doesn't matter if this column is used or not, ActiveRecord is caching the table columns. This is why it's crashing if you simply remove a column like you intended to.

Please remove the commented-out lines from the migration because this will be confusing anybody looking at the code.

dmarcoux replied:
The database schema is also wrong now since the column `hostsystem_id` is still being removed.

dmarcoux commented on src/api/db/schema.rb:
```diff
@@ -10,7 +10,7 @@
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2021_02_05_113812) do
+ActiveRecord::Schema.define(version: 2021_06_03_113812) do
```
This version number is off. How did you create the migration included in this PR? You edited its filename by hand, right? Because this timestamp is in the future, so it cannot be possible that you generated it with `bundle exec rails g migration MyMigration123`.

adrianschroeter commented:
On Montag, 3. Mai 2021, 17:03:11 CEST Dany Marcoux wrote:
> @dmarcoux commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,17 @@
> +class FixHostsystemDefinition < ActiveRecord::Migration[6.0]
> +  class SchemaMigration < ApplicationRecord
> +  end
> +
> +  def up
> +    # no real production data inside yet, so we skip any conversation.
> +    safety_assured { remove_column :repositories, :hostsystem_id }
> 
> The database schema is also wrong now since the column `hostsystem_id` is still being removed.

that is not important, as it should be created anyway and the content can not 
get used for anything so far.

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

coolo commented on src/api/app/models/project/update_from_xml_command.rb:
```diff
@@ -195,9 +195,22 @@ def update_repository_without_path_element(xml_hash)
+      position = 1
+      xml_hash.elements('hostsystem') do |hostsystem|
+        host_repo = Repository.find_by_project_and_name(hostsystem['project'], hostsystem['repository'])
+        if xml_hash['hostsystem']['project'] == project.name && xml_hash['hostsystem']['repository'] == xml_hash['name']
```
careful. You use `hostsystem` as `xml_hash['hostsystem']` and you should continue using that variable to access 'repository' and 'project'
Because  if 'hostsystem' can appear multiple times (as is the topic of your PR), then there is no single xml_hash['hostsystem']['project']

The fact that you didn't spot this yourself, shows a certain lack of test coverage

adrianschroeter replied:
yeah .. that is broken indeed ... (current code is even more broken though as it removes hostsystem even in other repositories)

adrianschroeter replied:
will add a simple test and a fix tomorrow morning, need to finish something else atm

adrianschroeter commented:
added a simple store and render test case now and fixed some more rubocop complains.

I will not touch it anymore except on request to avoid conflicts. But it would be good to get it deployed today, since it blocks the work on cross building projects atm.

hennevogel commented:
Pushed re-worked migrations, they are safe to run now. 

There is a spec for `Repository.hostsystem` though. Should this be removed?

adrianschroeter commented:
at least removed in that form, (we could re-introduce it in a form where it tests multiple hostsystem definitions in multiple repositories)

adrianschroeter commented:
did so now, thanks for updating the requests!

adrianschroeter commented:
assuming hennes push as review
